test/test-executor.c

```diff
@@ -0,0 +1,322 @@
+      if (data->head == data->tail) {
+        uv_mutex_unlock(&data->mutex);
+        break;
+      }
+
+      printf("worker: found %d work -- head %d tail %d\n", data->tail - data->head, data->head, data->tail);
+      ASSERT(0 <= data->head && data->head <= data->tail);
+      uv_mutex_unlock(&data->mutex);
+
+      /* Handle one. */
+      work = data->queued_work[data->head];
+      printf("worker: Running work %p\n", work);
+      work->work_cb(work);
+
+      /* Tell libuv we're done with this work. */
+      printf("worker: Telling libuv we're done with %p\n", work);
+      data->executor->done(work);
+
+      /* Advance. */
+      uv_mutex_lock(&data->mutex);
+      printf("worker: Advancing\n");
+      data->head++;
+      data->n_completed++;
+      uv_mutex_unlock(&data->mutex);
+    }
+
+    /* Loop unless (1) no work, and (2) no_more_work_coming. */
+    uv_mutex_lock(&data->mutex);
+    if (data->head == data->tail && data->no_more_work_coming) {
+      printf("worker exiting\n");
+      fflush(stdout);
+      should_break = 1;
+      uv_sem_post(&data->thread_exiting);
+    }
+    uv_mutex_unlock(&data->mutex);
+
+    if (should_break)
+      break;
+  }
+}
+
+static void toy_executor_init(uv_executor_t* executor) {
+  struct toy_executor_data* data;
+  
+  data = (struct toy_executor_data *) executor->data;
+  data->times_submit_called = 0;
+  data->times_cancel_called = 0;
+  data->n_completed = 0;
+  data->head = 0;
+  data->tail = 0;
+  data->no_more_work_coming = 0;
+  data->executor = executor;
+  ASSERT(0 == uv_mutex_init(&data->mutex));
+  ASSERT(0 == uv_sem_init(&data->thread_exiting, 0));
+
+  ASSERT(0 == uv_thread_create(&data->thread, worker, data));
+}
+
+static void toy_executor_destroy(uv_executor_t* executor) {
+  struct toy_executor_data* data;
+  
+  data = (struct toy_executor_data *) executor->data;
+  uv_thread_join(&data->thread);
+  uv_mutex_destroy(&data->mutex);
+}
+
+static void toy_executor_submit(uv_executor_t* executor,
+                                uv_work_t* req,
+                                const uv_work_options_t* opts) {
+  struct toy_executor_data* data;
+  printf("toy_executor_submit: req %p\n", req);
+  
+  data = (struct toy_executor_data *) executor->data;
+  data->times_submit_called++;
+
+  uv_mutex_lock(&data->mutex);
+  ASSERT(data->tail < ARRAY_SIZE(data->queued_work));
+  data->queued_work[data->tail] = req;
+  data->tail++;
+  uv_mutex_unlock(&data->mutex);
+}
+
+static int toy_executor_cancel(uv_executor_t* executor, uv_work_t* req) {
+  struct toy_executor_data* data;
+  
+  data = (struct toy_executor_data *) executor->data;
+  printf("toy_executor_cancel: req %p\n", req);
+  data->times_cancel_called++;
+
+  return UV_EINVAL;
+}
+
+static void toy_work(uv_work_t* req) {
+  printf("toy_work: req %p\n", req);
+  ASSERT(req);
+}
+
+static void toy_slow_work(uv_work_t* req) {
+  printf("toy_slow_work: req %p\n", req);
+  ASSERT(req);
+  uv_sem_wait((uv_sem_t *) req->data);
```

Comment:
Use `finish_slow_work` directly here?  That makes it clear what it's waiting on.